src/praisonai-agents/praisonaiagents/agent/durable.py

```
@@ -59,6 +59,52 @@ def _accepts_idempotency_key(execute_tool_fn: Callable) -> bool:
+    agent = getattr(executor, "__self__", None)
+    tools = getattr(agent, "tools", None)
+    if isinstance(tools, (list, tuple)):
+        for tool in tools:
+            name = getattr(tool, "name", None) or getattr(tool, "__name__", None)
+            if name == function_name and getattr(tool, "restart_safe", None) is not None:
+                return bool(tool.restart_safe)
+    try:
+        from ..tools import get_registry as _get_tool_registry
+
+        tool = _get_tool_registry().get(function_name)
+    except Exception:  # pragma: no cover - registry lookup is best-effort
```

Ruff (0.16.1)

[warning] Do not catch blind exception: `Exception`

(BLE001)